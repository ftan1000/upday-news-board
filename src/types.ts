// Note: Should be a dynamic list. Hard-coded here b/c of time-constraints
export const BoardList = {
	en: 'England',
	de: 'Deutsch',
	it: 'Italiano'
};

export const StatusList = {
	draft: 'Draft',
	published: 'Published',
	archived: 'Archived'
}