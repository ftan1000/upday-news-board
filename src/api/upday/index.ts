/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */
export { ApiError } from './core/ApiError';
export { CancelablePromise } from './core/CancelablePromise';
export { OpenAPI } from './core/OpenAPI';

export type { Board } from './models/Board';
export type { BoardNews } from './models/BoardNews';
export { News } from './models/News';

export { BoardService } from './services/BoardService';
export { NewsService } from './services/NewsService';
