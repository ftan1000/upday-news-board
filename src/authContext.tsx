import {createContext, ReactNode, useContext, useState} from 'react';

export type User = {
	email?: string
}

type authContextType = {
	user: User;
	login: (user: User) => void;
  refreshAuth: () => boolean;
	logout: () => void;
	isAuthenticated: boolean;
};

const authContextDefaultValues: authContextType = {
	user: {},
	login: () => {},
  refreshAuth: () => false,
	logout: () => {},
	isAuthenticated: false,
};

const AuthContext = createContext<authContextType>(authContextDefaultValues);

export function useAuth() {
	return useContext(AuthContext);
}

type Props = {
	children: ReactNode;
};

export function AuthProvider({children}: Props) {
	const [user, setUser] = useState<User>({});
	const [isAuthenticated, setAuthentication] = useState(false);

	const login = (user: User) => {
		setAuthentication(true);
		setUser({email: user.email});
    if (typeof window !== "undefined") {
      localStorage.setItem('email', user.email as string);
    }
	};

  const refreshAuth = (): boolean => {
    if (typeof window !== "undefined") {
      const email = localStorage.getItem('email');
      if (email) {
        login({email: email});
        return true;
      }
    }
    return false;
  };

	const logout = () => {
		setAuthentication(false);
		setUser({});
    if (typeof window !== "undefined") {
      localStorage.removeItem('email');
    }
	};

	const value = {
		user,
		login,
    refreshAuth: refreshAuth,
		logout,
		isAuthenticated,
	};

	return (
		<>
			<AuthContext.Provider value={value}>
				{children}
			</AuthContext.Provider>
		</>
	);
}
