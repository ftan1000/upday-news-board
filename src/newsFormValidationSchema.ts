import * as yup from 'yup';
import {AnySchema} from 'yup';
import {OptionalObjectSchema} from 'yup/lib/object';

const generateNewsFormValidationSchema = (): OptionalObjectSchema<Record<string, AnySchema>> =>
	yup.object({
		title: yup
			.string()
			.required('Title is required'),
		imageURL: yup
			.string()
			.url()
	});

export default generateNewsFormValidationSchema;
