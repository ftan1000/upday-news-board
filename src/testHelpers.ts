import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { HTMLAttributes } from 'react';

export const findByTestId = (wrapper: ShallowWrapper | ReactWrapper, testId: string): ReactWrapper<HTMLAttributes<any>, any> | ShallowWrapper<HTMLAttributes<any>, any> => {
  return wrapper.find(`[data-testid="${testId}"]`);
};
