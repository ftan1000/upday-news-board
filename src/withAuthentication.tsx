import { useRouter } from "next/router";
import { useEffect } from "react";
import type { FC } from "react";
import {useAuth} from './authContext';

type withAuthenticationFn = (Component: FC) => FC;

const withAuthentication: withAuthenticationFn = (Component) => {
	const Authenticated: FC = (): JSX.Element | null => {
		const { user } = useAuth();
		const router = useRouter();

		useEffect(() => {
			if (!user || !user.email) router.push("/login");
		});

		return user && user.email ? <Component /> : null;
	};

	return Authenticated;
};

export default withAuthentication;