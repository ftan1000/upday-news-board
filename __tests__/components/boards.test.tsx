import React from 'react';
import {shallow} from 'enzyme';
import {Board} from "../../src/api/upday";
import Boards from "../../components/boards";
import {findByTestId} from "../../src/testHelpers";
import {ListItemText} from "@mui/material";

describe('Boards', () => {

  const data: Board[] = [
      { id: 'en', name: 'England' },
      { id: 'de', name: 'Deutsch' },
      { id: 'it', name: 'Italiano' }
      ];

  const emptyData: Board[] = [];

  it('correctly renders ListItemText components when data exists', () => {

    const props = { data };
    const wrapper = shallow(<Boards {...props} />);

    expect(wrapper.find(ListItemText)).toHaveLength(data.length);
    for (let entry of data) {
      expect(findByTestId(wrapper, 'board-ListItemText-'+entry.id).exists()).toBeTruthy();
    }
    expect(findByTestId(wrapper, 'board-nodata-static-msg').exists()).toBeFalsy()
    expect(findByTestId(wrapper, 'board-nodata-error-msg').exists()).toBeFalsy();
  });

  it('renders static error msg when no data is given', () => {
    const wrapper = shallow(<Boards />);

    expect(wrapper.find(ListItemText)).toHaveLength(0);
    expect(findByTestId(wrapper, 'board-nodata-static-msg').exists()).toBeTruthy();
    expect(findByTestId(wrapper, 'board-nodata-error-msg').exists()).toBeFalsy();
  });

  it('renders static error msg when data is empty', () => {

    const props = { emptyData };

    // @ts-ignore
    const wrapper = shallow(<Boards {...props} />);

    expect(wrapper.find(ListItemText)).toHaveLength(0);
    expect(findByTestId(wrapper, 'board-nodata-static-msg').exists()).toBeTruthy();
    expect(findByTestId(wrapper, 'board-nodata-error-msg').exists()).toBeFalsy();
  });

  it('renders static and dynamic message when no data and error message are given', () => {

    const errorMessage = 'some-error-message';
    const props = { errorMessage: errorMessage, hasError:true, data: emptyData };
    const wrapper = shallow(<Boards {...props} />);

    expect(wrapper.find(ListItemText)).toHaveLength(0);
    expect(findByTestId(wrapper, 'board-nodata-static-msg').exists()).toBeTruthy();
    expect(findByTestId(wrapper, 'board-nodata-error-msg').exists()).toBeTruthy();
    expect(wrapper.contains(<pre>{errorMessage}</pre>)).toBeTruthy();
  });

});
