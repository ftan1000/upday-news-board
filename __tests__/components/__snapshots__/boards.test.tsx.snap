// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boards renders snapshot 1`] = `
Array [
  <h1>
    Boards
  </h1>,
  <div>
    Choose a board to view and manage the news:
    <ul
      className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
            data-testid="board-ListItemText-en"
            href="/board/en"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              England
            </span>
          </div>
        </div>
      </li>
      <hr
        className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
            data-testid="board-ListItemText-de"
            href="/board/de"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Deutsch
            </span>
          </div>
        </div>
      </li>
      <hr
        className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
            data-testid="board-ListItemText-it"
            href="/board/it"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Italiano
            </span>
          </div>
        </div>
      </li>
      <hr
        className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
    </ul>
  </div>,
]
`;
